icu4c/source/i18n/numfmt.cpp

```diff
@@ -948,7 +948,7 @@ class ICUNumberFormatService : public ICULocaleService {
     }
 
     virtual UObject* handleDefault(const ICUServiceKey& key, UnicodeString* /* actualID */, UErrorCode& status) const override {
-        LocaleKey& lkey = (LocaleKey&)key;
+        LocaleKey& lkey = dynamic_cast<LocaleKey&>(const_cast<ICUServiceKey&>(key));
```

Comment:
This particuular line, if change to use static_cast<>
get the following error during runtime under CFI
```
      IntlTestNumberFormatAPI {
         NumberFormat API test {
         } OK:   NumberFormat API test 
         NumberFormatRegistration {
numfmt.cpp:951:27: runtime error: control flow integrity check for type 'icu::LocaleKey' failed during base-to-derived cast (vtable address 0x7f25bf9fa730)
0x7f25bf9fa730: note: invalid vtable
 25 7f 00 00  a0 7e 8c bf 25 7f 00 00  80 7b 8c bf 25 7f 00 00  a0 7b 8c bf 25 7f 00 00  f0 59 8c bf
              ^ 
numfmt.cpp:951:27: note: check failed in ../../lib/libicui18n.so.73, vtable located in ../../lib/libicuuc.so.73
SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior numfmt.cpp:951:27 in
```